File: server/src/Ground43.Api/Infrastructure/RequirementSort.cs
namespace Ground43.Api.Infrastructure;

internal static class RequirementSort
{
    public static int StatusRank(string id) => id switch
    { "todo" => 0, "in_progress" => 1, "review" => 2, "paused" => 3, "backlog" => 4, "completed" => 5, "closed" => 6, _ => 7 };
    public static int PriorityRank(string priority) => priority switch
    { "urgent" => 0, "high" => 1, "medium" => 2, _ => 3 };
    public static long Number(string id) => long.TryParse(id.StartsWith("REQ-", StringComparison.OrdinalIgnoreCase) ? id[4..] : id, out var number) ? number : long.MaxValue;
}

File: client/src/requirementSort.ts
import type { Requirement } from './types'
import { priorityRank } from './utils'

const statusOrder = ['todo', 'in_progress', 'review', 'paused', 'backlog', 'completed', 'closed']
export function compareRequirementNumber(a: string, b: string) {
  return a.localeCompare(b, 'en', { numeric: true })
}
export function comparePriorityAndNumber(a: Requirement, b: Requirement) {
  return priorityRank[b.priority] - priorityRank[a.priority] || compareRequirementNumber(a.id, b.id)
}
export function compareTaskStatus(a: Requirement, b: Requirement) {
  const rank = (id: string) => { const index = statusOrder.indexOf(id); return index < 0 ? statusOrder.length : index }
  return rank(a.statusId) - rank(b.statusId) || comparePriorityAndNumber(a, b)
}
